services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -64,31 +65,38 @@
 import java.util.Queue;
 import java.util.Set;
 
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
 /**
  * Provides a reflection service for Protobuf services (including the reflection service itself).
  *
- * <p>Throws an exception if the set of services contain multiple protobuf files with declarations
- * of the same service, method, type, or extension.
+ * <p>Separately tracks mutable and immutable services. Throws an exception if either group of
+ * services contains multiple Protobuf files with declarations of the same service, method, type, or
+ * extension.
  */
 @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
 public final class ProtoReflectionService extends ServerReflectionGrpc.ServerReflectionImplBase
```

Comment:
The only advantage given in Effective Java's Item 1 for a static factory method that applies to a constructor without parameters is the ability to return a subtype of the return type. The return type needs to implement `BindableService` and `InternalNotifyOnServerBuild` to work properly here, so a factory method can either return `ProtoReflectionService` (invalidating the advantage of the factory method) or a type that implements both `BindableService` and `InternalNotifyOnServerBuild`, which at the moment does not exist.

Now that I write this, it may be reasonable for `InternalNotifyOnServerBuild` to itself extend `BindableService`. But then the new `getInstance()` static method will return a type with `Internal` in its name, and I'm not sure that's a benefit from an API commitment perspective.

Anyways, I don't think this really has anything to do with more efficient handling of dynamic services. I think that `InternalNotifyOnServerBuild` should probably extend `BindableService` (it needs to in order to do anything useful internally anyways), but this and the decision of replacing the constructor with a static factory method can probably be separated from this PR.

(Side note: I believe it's possible to start multiple servers in the same program, in which case it would be valid to have multiple instances of the reflection service)